### PR TITLE
25 installations by company size

### LINKED ISSUE #25: Explore installations by company size

### PR DESCRIPTION
**Pull request by crispy-wonton**
Fixes #25 

**Description**

This notebook contains analysis to understand the counts and proportions of heat pumps installed by differently sized companies. We have used the MCS installations data, using annual installations per company, to produce a proxy for this. The research questions to answer are as follows:
1. Which companies are the top 30 installers in Q1 2025 by installation count?
2. Which companies are the top 10 installers in 2024 by installation count?
3. What are the proportions of installations by differently sized companies over time?
4. What are the counts of installations by differently sized companies over time?
5. What is the breakdown of the number of companies per size bracket in 2024?

**Instructions for reviewer:**

**To run the code:**
- Checkout the branch and create a new environment, activate it, then install `requirements.txt`
- run `jupytext --to notebook installations_by_company_size/installations_by_company_size.py` to create the notebook
- run the notebook

**Please pay special attention to:**
- Please could you check I have applied manipulations correctly to answer the questions.
- Please could you let me know if you think there are other data processing steps that should happen before the analysis. I also plan to double check this with Sofia next week as she knows the MCS data the best.

### CODE REVIEW

File: installations_by_company_size/installations_by_company_size.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+# %%
+installations_df = pl.read_parquet("s3://asf-daps/lakehouse/2025_Q1/processed/mcs/mcs_installations_250611-0.parquet")
+
+# %%
+installations_df = installations_df.drop("__index_level_0__").filter(
+    # Check if we want to include any other installation_type fields
+    pl.col("installation_type") == "Domestic",
+    pl.col("tech_type") != "Micro CHP",
+).with_columns(
+    # Drop duplicate rows - check that we should do this. Could these represent true multi-installs?
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py:30]: This isn't dropping anything. Not too worrying, but note there are ~2k duplicates if just looking at addresses. Could just see if your results change using that, although it of course risks dropping legitimate multiple installs as you flag.

File: installations_by_company_size/installations_by_company_size.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+# %%
+import polars as pl
+import matplotlib.pyplot as plt
+import numpy as np
+
+# %%
+installations_df = pl.read_parquet("s3://asf-daps/lakehouse/2025_Q1/processed/mcs/mcs_installations_250611-0.parquet")
+
+# %%
+installations_df = installations_df.drop("__index_level_0__").filter(
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: Don't save back to a data frame of the same name.

File: installations_by_company_size/installations_by_company_size.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    "1-10": 0,
+    "11-50": 1,
+    "51-100": 2,
+    "101-365": 3,
+    "Over 365": 4
+}
+
+# %%
+# Aggregate the number of installations per installer company per year
+installations_per_installer_per_year = installations_df.group_by(["company_unique_id", "commission_year"]).len().rename({"len": "n_installations"}).with_columns(
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: Use pl.cut().
```
range_labels = ["1-10", "11-50", "51-100", "101-365", "Over 365"]
range_breaks = [10, 50, 100, 365]

installations_per_installer_per_year = (
    installations_df.group_by(["company_unique_id", "commission_year"])
    .len()
    .rename({"len": "n_installations"})
    .with_columns(
        (pl.col("n_installations") / pl.col("n_installations").sum().over("commission_year"))
        .alias("proportion_of_total_installations_per_year"),
        pl.col("n_installations").cut(breaks=range_breaks, labels=range_labels).alias("n_installations_range"),
    )
    .with_columns(
        pl.col("n_installations_range").replace(order_of_ranges).cast(pl.Int8).alias("range_order")
    )
    .sort("range_order")
)
```

File: installations_by_company_size/installations_by_company_size.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    pl.when(pl.col("n_installations") <= 10)
+    .then(pl.lit("1-10"))
+    # Both ends inclusive
+    .when(pl.col("n_installations").is_between(11, 50, closed="both"))
+    .then(pl.lit("11-50"))
+    .when(pl.col("n_installations").is_between(51, 100, closed="both"))
+    .then(pl.lit("51-100"))
+    .when(pl.col("n_installations").is_between(101, 365, closed="both"))
+    .then(pl.lit("101-365"))
+    .when(pl.col("n_installations") > 365)
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: I find this so confusing lol, keep thinking it's dates because 365 is being used.

File: installations_by_company_size/installations_by_company_size.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    pl.col("installation_type") == "Domestic",
+    pl.col("tech_type") != "Micro CHP",
+).with_columns(
+    # Drop duplicate rows - check that we should do this. Could these represent true multi-installs?
+    pl.col(pl.String).str.to_lowercase()
+).unique()
+
+# %%
+# This is to ensure the ranges show in the desired order on the plots
+order_of_ranges = {
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py:37]: Any particular logic for these buckets? Choice of buckets will definitely dictate what we visualise here.

File: installations_by_company_size/installations_by_company_size.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    pl.col("n_installations").sum(),
+    pl.col("range_order").first(),
+    pl.col("company_unique_id").n_unique().alias("n_companies")
+).with_columns(
+    (pl.col("n_installations") / pl.col("n_installations").sum().over(["commission_year"])).alias("proportion_of_total_installations_per_year")
+).sort(by=["range_order", "commission_year"])
+
+# %%
+# Show the proportions of installations by each group over time
+commission_years = installations_per_size_per_year["commission_year"].unique().sort()
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: Might consider dropping earlier years (maybe earlier than 2014?) when we have very few instllations. I think only 183 records from 2014 and before!

File: installations_by_company_size/installations_by_company_size.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    cm = plt.get_cmap(cmap)
+    pts = np.linspace(0, 1, n)
+    return [cm(pt) for pt in pts]
+
+colours = create_colour_map("viridis", len(order_of_ranges))
+
+fig = plt.figure(figsize=(12, 5))
+
+# Plot each time series
+for r, colour in zip(order_of_ranges, colours):
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py:107]: This might be much more nicely visualised as an area chart.

Comment (caldwellst) [reply installations_by_company_size/installations_by_company_size.py:107]:
```
# Show proportions of installations by each group over time
installations_per_size_per_year_from_2015 = installations_per_size_per_year.filter(pl.col("commission_year") >= 2015)

# colour_map = {r: colour for r in order_of_ranges}
def create_colour_map(cmap, n):
    cm = plt.get_cmap(cmap)
    pts = np.linspace(0, 1, n)
    return [cm(pt) for pt in pts]

colours = create_colour_map("viridis", len(order_of_ranges))

proportions_wide = (
    installations_per_size_per_year_from_2015
    .pivot(
        on="n_installations_range",
        index="commission_year",
        values="proportion_of_total_installations_per_year",
    )
    .fill_null(0)
    .sort("commission_year")
)

commission_years = proportions_wide["commission_year"].to_numpy()
stacked_proportions = [proportions_wide[r].to_numpy() for r in order_of_ranges]

fig = plt.figure(figsize=(12, 5))

plt.stackplot(
    commission_years,
    stacked_proportions,
    colors=colours,
    labels=list(order_of_ranges),
)

plt.title("Proportion of annual domestic heat pump installations by company installation count")
plt.xlabel("Year")
plt.ylabel("Proportion of total annual installations")
plt.legend(title="Number of annual installations per company*", loc="upper left", bbox_to_anchor=(1.02, 1))
plt.xticks(commission_years)
plt.xlim(commission_years.min(), commission_years.max())
plt.ylim(0, 1)

text = "*The groups represent the number of installations per year per company. " \
"The lines in the plot will therefore represent different total numbers of companies in different years as the number of installations by individual companies grows."
plt.figtext(
    0.5, -0.06, text, wrap=True, horizontalalignment="center"
)
```

File: installations_by_company_size/installations_by_company_size.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+fig = plt.figure(figsize=(12, 5))
+
+for r, colour in zip(order_of_ranges, colours):
+    plot_data = installations_per_size_per_year.filter(pl.col("n_installations_range") == r)
+    plt.plot(plot_data["commission_year"], plot_data["n_companies"], color=colour, label=r)
+
+plt.title("Count of installer companies by number of annual installations per year")
+plt.xlabel("Year")
+plt.ylabel("Count of installer companies")
+plt.legend(title="Number of annual installations per company*")
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: You don't have the asterisk caption on this plot.

File: installations_by_company_size/installations_by_company_size.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+plt.xlabel("Year")
+plt.ylabel("Count of total annual installations")
+plt.legend(title="Number of annual installations per company*")
+plt.xticks(commission_years)
+
+text = "*The groups represent the number of installations per year per company. " \
+"The lines in the plot will therefore represent different total numbers of companies in different years as the number of installations by individual companies grows."
+plt.figtext(
+    0.5, -0.06, text, wrap=True, horizontalalignment="center"
+)
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: Similarly might be nicer as an area plot? Shows the stacked/cumulative effect better.

```
# Show count of installations over time
counts_wide = (
    installations_per_size_per_year
    .filter(pl.col("commission_year") >= 2015)
    .pivot(
        on="n_installations_range",
        index="commission_year",
        values="n_installations",
    )
    .fill_null(0)
    .sort("commission_year")
)

commission_years = counts_wide["commission_year"].to_numpy()
stacked_counts = [counts_wide[r].to_numpy() for r in order_of_ranges]

colours = create_colour_map("viridis", len(order_of_ranges))

fig = plt.figure(figsize=(12, 5))

plt.stackplot(
    commission_years,
    stacked_counts,
    colors=colours,
    labels=list(order_of_ranges),
)

plt.title("Count of annual domestic heat pump installations by company installation count")
plt.xlabel("Year")
plt.ylabel("Count of total annual installations")
plt.legend(title="Number of annual installations per company*", loc="upper left", bbox_to_anchor=(1.02, 1))
plt.xticks(commission_years)
plt.xlim(commission_years.min(), commission_years.max())

text = "*The groups represent the number of installations per year per company. " \
"The areas in the plot will therefore represent different total numbers of companies in different years as the number of installations by individual companies grows."
plt.figtext(
    0.5, -0.06, text, wrap=True, horizontalalignment="center"
)
```

File: installations_by_company_size/installations_by_company_size.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+plt.xlabel("Year")
+plt.ylabel("Count of installer companies")
+plt.legend(title="Number of annual installations per company*")
+plt.xticks(commission_years)
+plt.show()
+
+# %%
+# Breakdown of companies per bracket in 2024
+installations_per_size_per_year.filter(pl.col("commission_year") == 2024).with_columns(
+    (pl.col("n_companies") / pl.col("n_companies").sum() * 100).alias("pc_of_companies")
```
Comment (caldwellst) [inline installations_by_company_size/installations_by_company_size.py]: Maybe also just add in pc_of_installations for each row as well.

File: installations_by_company_size/installations_by_company_size.py (hunk 3 above)
Comment (caldwellst) [reply installations_by_company_size/installations_by_company_size.py]: How confident are we in the "uniqueness" of the strings for the company ID? I assume that is covered in DAPS, so I don't have a sense here, but would want us to be confident in that.

**Review by caldwellst — CHANGES_REQUESTED** [review]
Have added my comments. A couple of thoughts on how we're filtering out the data, but overall the filtering looks okay to me, but trust y'all and your DAPS processing, so check with Sofia when she returns.

A couple of thoughts on how you visualise, as I think area plots more relevant to show stacked/cumulative effects! Overall, looks good and maybe worth getting something over to MCS for approval sooner rather than later.

File: installations_by_company_size/installations_by_company_size.py (hunk 5 above)
Comment (crispy-wonton) [reply installations_by_company_size/installations_by_company_size.py:37]: I had a look at the distributions and came up with these somewhat arbitrarily, then sense-checked with a couple of people in the team and they were happy to go ahead with them.

File: installations_by_company_size/installations_by_company_size.py (hunk 3 above)
Comment (crispy-wonton) [reply installations_by_company_size/installations_by_company_size.py]: Thank you for the `pl.cut()`!! Much better.

On your q about uniqueness - yeh I don't know so that's something I want to check with Sofia next week. There's too many unique names to manually check but if it's not covered in daps I could add a basic similarity check here.

File: installations_by_company_size/installations_by_company_size.py (hunk 1 above)
Comment (crispy-wonton) [reply installations_by_company_size/installations_by_company_size.py:30]: Yeh, I kept the line for now in case we wanted to update it with more specific deduplication. I also tested it with address and I thought that seemed like a high number of duplicates.

This is something to check with Sofia as I think we've discussed the duplicates with MCS before.

File: installations_by_company_size/installations_by_company_size.py (hunk 7 above)
Comment (crispy-wonton) [reply installations_by_company_size/installations_by_company_size.py:107]: oooh I love it!

File: installations_by_company_size/installations_by_company_size.py (hunk 1 above)
Comment (crispy-wonton) [reply installations_by_company_size/installations_by_company_size.py:30]: I've confirmed with Sofia that there's no deduplication required and that any duplicate addresses should represent genuine records.

File: installations_by_company_size/installations_by_company_size.py (hunk 3 above)
Comment (crispy-wonton) [reply installations_by_company_size/installations_by_company_size.py]: I have checked with Sofia about this.

The company unique ID should represent unique companies. It is created in our internal preprocessing to aggregate some companies which appear under slightly different strings in the `installer_name` column through a standardisation process. It's possible there are still some companies that appear under slightly different names here but it sounds like sufficient effort has been made to standardise the names.
